RockWeb/Webhooks/WebhookToWorkflow.ashx

```diff
@@ -0,0 +1,394 @@
+    /// <summary>
+    /// Convert the request into a generic JSON object that can provide information
+    /// to the workflow. If a subclass does needs to customize this data they can
+    /// call the base method and then modify the content before returning it.
+    /// </summary>
+    /// <param name="hook">The DefinedValue of the currently executing webhook.</param>
+    /// <returns></returns>
+    protected virtual Dictionary<string, object> RequestToDictionary( DefinedValueCache hook )
+    {
+        var dictionary = new Dictionary<string, object>();
+
+        //
+        // Set the standard values to be used.
+        //
+        dictionary.Add( "DefinedValueId", hook.Id );
+        dictionary.Add( "Url", Url );
+        dictionary.Add( "RawUrl", HttpContext.Request.Url.AbsoluteUri );
+        dictionary.Add( "Method", HttpContext.Request.HttpMethod );
+        dictionary.Add( "QueryString", HttpContext.Request.QueryString.Cast<string>().ToDictionary( q => q, q => HttpContext.Request.QueryString[q] ) );
+        dictionary.Add( "RemoteAddress", HttpContext.Request.UserHostAddress );
+        dictionary.Add( "RemoteName", HttpContext.Request.UserHostName );
+        dictionary.Add( "ServerName", HttpContext.Request.Url.Host );
+
+        //
+        // Add in the raw body content.
+        //
+        using ( StreamReader reader = new StreamReader( HttpContext.Request.InputStream, Encoding.UTF8 ) )
+        {
+            dictionary.Add( "RawBody", reader.ReadToEnd() );
+        }
+
+        //
+        // Parse the body content if it is JSON or standard Form data.
+        //
+        if ( HttpContext.Request.ContentType == "application/json" )
+        {
+            try
+            {
+                dictionary.Add( "Body", Newtonsoft.Json.JsonConvert.DeserializeObject( ( string ) dictionary["RawBody"] ) );
+            }
+            catch
+            {
+            }
+        }
+        else if ( HttpContext.Request.ContentType == "application/x-www-form-urlencoded" )
+        {
+            try
+            {
+                dictionary.Add( "Body", HttpContext.Request.Form.Cast<string>().ToDictionary( q => q, q => HttpContext.Request.Form[q] ) );
+            }
+            catch
+            {
+            }
+        }
+        else if ( HttpContext.Request.ContentType == "application/xml" )
+        {
+            try
+            {
+                XmlDocument doc = new XmlDocument();
+                doc.LoadXml( ( string ) dictionary["RawBody"] );
+                string jsonText = JsonConvert.SerializeXmlNode( doc );
+                dictionary.Add( "Body", Newtonsoft.Json.JsonConvert.DeserializeObject( ( jsonText ) ) );
+            }
+            catch
+            {
+            }
+        }
+
+        //
+        // Add in all the headers if the admin wants them.
+        //
+        if ( hook.GetAttributeValue( "Headers" ).AsBoolean() )
+        {
+            var headers = HttpContext.Request.Headers.Cast<string>()
```

Comment:
When I originally wrote this code I made those optional since they can include security related information. For example, Basic HTTP Auth headers and cookies that allow somebody to impersonate (potentially). The reason for making it optional was because everything was going to be automatically stuffed into a `Request` attribute that normal users could then potentially see.

If I understand all your other comments this won't be happening anymore and the admin will be picking and choosing what attributes get populated with what data _only_, so there shouldn't be a security concern anymore and indeed those "include headers" and "include cookies" options aren't needed.